File: hs/Util/ParseEngine.hs

module Util.ParseEngine
   (
   next5x1,
   next6x1,
   next7x1,
   next8x1,
   next9x1,
   next5x2,
   next6x2,
   next7x2,
   next8x2,
   next9x2,
   Rule, Table,
   parse,
   module Stream
   ) where

import qualified Data.ByteString as B
import Data.List
import Control.Exception
import Control.Monad
import Array
import Util.Stream as Stream

next5x1 :: B.ByteString -> Int -> Int -> Int
next5x1 table state symbol =
   let i = 32 * state + symbol
   in
      fromIntegral (B.index table i) - 128

next6x1 :: B.ByteString -> Int -> Int -> Int
next6x1 table state symbol =
   let i = 64 * state + symbol
   in
      fromIntegral (B.index table i) - 128

next7x1 :: B.ByteString -> Int -> Int -> Int
next7x1 table state symbol =
   let i = 128 * state + symbol
   in
      fromIntegral (B.index table i) - 128

next8x1 :: B.ByteString -> Int -> Int -> Int
next8x1 table state symbol =
   let i = 256 * state + symbol
   in
      fromIntegral (B.index table i) - 128

next9x1 :: B.ByteString -> Int -> Int -> Int
next9x1 table state symbol =
   let i = 512 * state + symbol
   in
      fromIntegral (B.index table i) - 128

next5x2 :: B.ByteString -> Int -> Int -> Int
next5x2 table state symbol =
   let i = 64 * state + 2 * symbol
   in
      fromIntegral (B.index table i) * 256 +
      fromIntegral (B.index table (i+1))
      - 32768

next6x2 :: B.ByteString -> Int -> Int -> Int
next6x2 table state symbol =
   let i = 128 * state + 2 * symbol
   in
      fromIntegral (B.index table i) * 256 +
      fromIntegral (B.index table (i+1))
      - 32768

next7x2 :: B.ByteString -> Int -> Int -> Int
next7x2 table state symbol =
   let i = 256 * state + 2 * symbol
   in
      fromIntegral (B.index table i) * 256 +
      fromIntegral (B.index table (i+1))
      - 32768

next8x2 :: B.ByteString -> Int -> Int -> Int
next8x2 table state symbol =
   let i = 512 * state + 2 * symbol
   in
      fromIntegral (B.index table i) * 256 +
      fromIntegral (B.index table (i+1))
      - 32768

next9x2 :: B.ByteString -> Int -> Int -> Int
next9x2 table state symbol =
   let i = 1024 * state + 2 * symbol
   in
      fromIntegral (B.index table i) * 256 +
      fromIntegral (B.index table (i+1))
      - 32768

type Rule value = (Int, Int, [value] -> [value])  {- lhs, size of rhs, function to call -}

type Table s m terminal value a =
   (terminal -> (Int, value),       {- reader -}
    Int -> Int -> Int,              {- action table -}
    Int -> Int -> Int,              {- goto table -}
    Array Int (Rule value),         {- reduction information -}
    value -> a,                     {- result destructor -}
    s terminal -> m SomeException)  {- error function -}

parse :: Streamable s m => Table s m terminal value a -> s terminal -> m (a, s terminal)
parse (read, action, goto, reduceTable, destruct, error) s =
   let
      {
      loop ststack valstack s =
         do {
            fr <- front s;
            case fr of
               {
               Nil ->
                  loopRead ststack valstack 0 undefined s s ;

               Cons term s' ->
                  let { (ordinal, value) = read term }
                  in
                     loopRead ststack valstack ordinal value s s'
               }
            } ;

      loopRead ststack valstack ordinal value s s' =
         case ststack of
            {
            [] ->
               fail "bad parsing table" ;

            state : rest ->
               let { n = action state ordinal }
               in
                  if n == 0 then
                     error s >>= throw
                  else if n > 0 then
                     {- shift -}
                     loop (n-1 : ststack) (value : valstack) s'
                  else if n == -1 then
                     {- accept -}
                     return (valstack, s)
                  else
                     {- reduce -}
                     let
                        {
                        (lhs, rhssize, f) = (Array.!) reduceTable (-n - 2) ;
                        ststack' = drop rhssize ststack
                        }
                     in
                        loopRead (goto (head ststack') lhs : ststack') (f valstack) ordinal value s s'
            }
      }
   in do
      {
      (valstack, s') <- loop [0] [] s;
      case valstack of
         {
         [value] ->
            return (destruct value, s') ;

         _ ->
            fail "bad parsing table"
         }
      }

File: ocaml/ParseEngine.ml

module type S =
   sig
   
      module Strm : Streamable.S

      type terminal
      type value

      val next5x1 : string -> int -> int -> int
      val next6x1 : string -> int -> int -> int
      val next7x1 : string -> int -> int -> int
      val next8x1 : string -> int -> int -> int
      val next9x1 : string -> int -> int -> int
      val next5x2 : string -> int -> int -> int
      val next6x2 : string -> int -> int -> int
      val next7x2 : string -> int -> int -> int
      val next8x2 : string -> int -> int -> int
      val next9x2 : string -> int -> int -> int

      type action = value list -> value list

      type 'a table =
         (int -> int -> int)             (* action table *)
         *
         (int -> int -> int)             (* goto table *)
         *
         (int * int * action) array      (* reduction information: lhs ordinal, rhs size, function *)
         *
         (value -> 'a)                   (* result destructor *)
         *
         (terminal Strm.t -> exn)        (* error function *)

      val parse : 'a table -> terminal Strm.t -> 'a * terminal Strm.t

   end


module Make (M : sig
                    module Strm : Streamable.S
                    type terminal
                    type value
                    val dummy : value
                    val read : terminal -> int * value
                 end)
   : S with type 'a Strm.t = 'a M.Strm.t
       and type terminal = M.terminal
       and type value = M.value
   =
   struct

      module Strm = M.Strm

      type terminal = M.terminal
      type value = M.value

      (* Transition function for a 5-bit symbols, with 1-byte results. *)
      let next5x1 table state symbol =
         let i = 32 * state + symbol
         in                       
            int_of_char (String.get table i) - 128

      let next6x1 table state symbol =
         let i = 64 * state + symbol
         in                       
            int_of_char (String.get table i) - 128

      let next7x1 table state symbol =
         let i = 128 * state + symbol
         in                       
            int_of_char (String.get table i) - 128

      let next8x1 table state symbol =
         let i = 256 * state + symbol
         in                       
            int_of_char (String.get table i) - 128

      let next9x1 table state symbol =
         let i = 512 * state + symbol
         in                       
            int_of_char (String.get table i) - 128

      let next5x2 table state symbol =
         let i = 64 * state + 2 * symbol
         in                       
            int_of_char (String.get table i) * 256 +
            int_of_char (String.get table (i+1))
            - 32768

      let next6x2 table state symbol =
         let i = 128 * state + 2 * symbol
         in                       
            int_of_char (String.get table i) * 256 +
            int_of_char (String.get table (i+1))
            - 32768

      let next7x2 table state symbol =
         let i = 256 * state + 2 * symbol
         in                       
            int_of_char (String.get table i) * 256 +
            int_of_char (String.get table (i+1))
            - 32768

      let next8x2 table state symbol =
         let i = 512 * state + 2 * symbol
         in                       
            int_of_char (String.get table i) * 256 +
            int_of_char (String.get table (i+1))
            - 32768

      let next9x2 table state symbol =
         let i = 1024 * state + 2 * symbol
         in                       
            int_of_char (String.get table i) * 256 +
            int_of_char (String.get table (i+1))
            - 32768


      type action = value list -> value list

      type 'a table =
         (int -> int -> int)             (* action table *)
         *
         (int -> int -> int)             (* goto table *)
         *
         (int * int * action) array      (* reduction information: lhs, size of rhs, letctions to call *)
         *
         (value -> 'a)                   (* result destructor *)
         *
         (terminal Strm.t -> exn)        (* error letction *)

      let parse (action, goto, reduce, destruct, error) s =
         let rec loop ststack valstack s =
            (match Strm.front s with
                Strm.Nil ->
                   loopRead ststack valstack 0 M.dummy s s
              | Strm.Cons (term, s') ->
                   let (ordinal, value) = M.read term
                   in
                      loopRead ststack valstack ordinal value s s')
            
         and loopRead ststack valstack ordinal value s s' =
            (match ststack with
                [] ->
                   raise (Failure "bad parsing table")
              | state :: rest ->
                   let n = action state ordinal
                   in
                      if n = 0 then
                         raise (error s)
                      else if n > 0 then
                         (* shift *)
                         loop (n-1 :: ststack) (value :: valstack) s'
                      else if n = -1 then
                         (* accept *)
                         (valstack, s)
                      else
                         (* reduce *)
                         let (lhs, rhssize, f) = Array.get reduce (-n - 2)
                         in
                         let ststack' = ListUtil.drop ststack rhssize
                         in
                            loopRead (goto (List.hd ststack') lhs :: ststack') (f valstack) ordinal value s s')

         in
            (match loop [0] [] s with
                ([value], s') -> (destruct value, s')
              | _ ->
                   raise (Failure "bad parsing table"))

   end
